### PR TITLE
[Ready for review] Boris Bike story

### LINKED ISSUE #172: [Turing Data Story] Boris Bike analysis
# Story description 

TfL publishes usage data for Boris Bikes: https://cycling.data.tfl.gov.uk/ It’s actually really detailed, to the point where I wonder if it might have similar privacy issues as NYC’s infamous taxi data: It shows each individual journey, in the format of “bike number X left bike station Y at time Z and arrived at station V at time W”.

I've done in the past an analysis where I tried to forecast how many bikes would be leaving/arriving at a given station, based on historical trends and weather data: http://nbviewer.org/github/mhauru/boris-bike-forecast/blob/master/analysis.ipynb We could base the story on that or take some entirely different angle, there's plenty you could do with the data given how detailed it is (most popular routes, popularity of routes as a function of time, effect of events or infrastructure changes on Boris Bike usage, all sorts of map visualisations, etc etc.). Proposals welcome.

# Ethical guideline 

Ideally a Turing Data Story has these properties and follows the [5 safes framework](http://blog.ukdataservice.ac.uk/access-to-sensitive-data-for-research-the-5-safes/).

- [ ] The analysis you produce is openly available and reproducible.
- [ ] Any data used are [open](https://opendatahandbook.org/guide/en/what-is-open-data/) and have an explicit licence, provenance and attribution.
- [ ] Any data used are not [personal data](https://ico.org.uk/for-organisations/guide-to-data-protection/guide-to-the-general-data-protection-regulation-gdpr/key-definitions/what-is-personal-data/) (i.e. the data is anonymous or anonymised).
- [ ] Any linkage of datasets in your data story does not lead to an increased risk of the personal identification of individuals.
- [ ] The Story must be truthful and clear about any limitations of analysis (and potential biases in data).
- [ ] The Story will not lead to negative social outcomes, such as (but not limited to) increasing discrimination or injustice.

# Current status

 - [x] Scoping of the story
 - [ ] Write story outline
 - [ ] Add material to the story notebook
 - [ ] Combine materials into a readable story
 - [ ] Proofread
 - [ ] Request reviews
 - [ ] Address reviews
 - [ ] Merge to master branch.

### PR DESCRIPTION
**Pull request by crangelsmith**
### Summary

Fixes #172

### List of changes proposed in this PR (pull-request)

* Notebook with Boris bike story

### What should a reviewer concentrate their feedback on?

- [ ] Review guidelines in issue 222
- [ ] Everything looks ok?

### Acknowledging contributors

- [ ] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [ ] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file: 
    - James Bishop and Federico Nanni

### CODE REVIEW

**Comment by llewelld** [comment]
<p>Please ignore this comment. After trying ot out I see that <code>wget</code> works rather nicely in the notebook, so better to leave this as it is.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506448'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by rchan26** [comment]
<p>Agree with David - I think an overview of the analysis will be nice (not necessarily the results, but the questions asked)</p><p> </p><p>I like the context and explanation of Boris bikes, but maybe worth a mention of other schemes: don't need to list them, but think worth saying alternatives exist like Lime, etc. but haven't been around as long and don't have the same docking mechanism that Boris bikes have. Only a sentence or two</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506335'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by rchan26** [comment]
<p>Should mention the units used for Duration (seems obvious its minutes, but can be made clear)</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506378'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by rchan26** [comment]
<blockquote>if you enjoy boredom</blockquote><p>I found this funny (as I do find it boring), but maybe someone who does like data cleaning wouldn't haha, but happy to allow this</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506397'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by rchan26** [comment]
<p>also, is the comment wrong? aren't you dropping entries where start date is _after_ end date?</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506513'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by rchan26** [comment]
<p>I also don't understand how it's greater than 1</p><p> </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506799'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by rchan26** [comment]
<p>agree with David - I'm a bit confused to what the jump is as I thought a jump would break a chain?</p><p> </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506992'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by rchan26** [comment]
<p>as with David, I've enjoyed reading this a lot! good job!</p><p> </p><p>I will look more closely to the code when I attempt to run it!</p><p> </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517507041'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: binder/environment.yml [hunk 1]
```diff
@@ -1,26 +1,201 @@
...
-  - scipy
-  - matplotlib
-  - seaborn
-  - openpyxl
-  - jupyter
-  - autopep8
-  - conda-forge::jupyter_contrib_nbextensions
-  - pymc3
-  - scikit-learn
+  - brotli=1.0.9
```
Comment (llewelld) [inline binder/environment.yml:6]: In order to get the conda environment to work on an m1 mac I had to make the following changes:
```diff
diff --git a/binder/environment.yml b/binder/environment.yml
index 726920e..7705069 100644
--- a/binder/environment.yml
+++ b/binder/environment.yml
@@ -51 +50,0 @@ dependencies:
-  - python=3.9.13
@@ -53 +51,0 @@ dependencies:
-  - python_abi=3.9
@@ -95 +93 @@ dependencies:
-    - fiona==1.8.21
+    - fiona>=1.8.21
@@ -125 +123 @@ dependencies:
-    - matplotlib==3.5.2
+    - matplotlib>=3.5.2
@@ -157 +155 @@ dependencies:
-    - pyproj==3.3.1
+    - pyproj>=3.3.1
@@ -164,2 +162,2 @@ dependencies:
-    - pyzmq==23.2.0
-    - rasterio==1.3.2
+    - pyzmq>=23.2.0
+    - rasterio>=1.3.2
@@ -169 +167 @@ dependencies:
-    - ruamel-yaml-clib==0.2.6
+    - ruamel-yaml-clib==0.2.7
```

File: stories/2022-10-06-Boris-Bikes/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,134 @@
+anyio==3.6.1; python_full_version >= "3.6.2" and python_version >= "3.7"
```
Comment (llewelld) [inline stories/2022-10-06-Boris-Bikes/requirements.txt:1]: In order to get the notebook to work with pip I had to make the following changes:
```diff
diff --git a/stories/2022-10-06-Boris-Bikes/requirements.txt b/stories/2022-10-06-Boris-Bikes/requirements.txt
index 5fb5347..492a415 100644
--- a/stories/2022-10-06-Boris-Bikes/requirements.txt
+++ b/stories/2022-10-06-Boris-Bikes/requirements.txt
@@ -30 +30 @@ fastjsonschema==2.16.1; python_version >= "3.7"
-fiona==1.8.21; python_version >= "3.8"
+fiona>=1.8.21; python_version >= "3.8"
@@ -61 +61 @@ matplotlib-inline==0.1.3; python_version >= "3.8"
-matplotlib==3.5.2; python_version >= "3.7"
+matplotlib>=3.5.2; python_version >= "3.7"
@@ -95 +95 @@ pyparsing==3.0.9; python_full_version >= "3.6.8" and python_version >= "3.7"
-pyproj==3.3.1; python_version >= "3.8"
+pyproj>=3.3.1; python_version >= "3.8"
@@ -105 +105 @@ pyyaml==6.0; python_version >= "3.6"
-pyzmq==23.2.0; python_version >= "3.7"
+pyzmq>=23.2.0; python_version >= "3.7"
```

**Review by llewelld — COMMENTED** [review]
The notebook ran locally nicely on an m1 mac after making just a few changes to the environment and requirements files as shown below. I also had to install a few other non-python dependencies manually:
```
brew install geos
brew install spatialindex
brew install wget
```

**Comment by crangelsmith** [comment]
<p>done, we've added some text.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506356'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>added licence information and units and shifted the rows.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506378'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>made things consistent, but we really want to leave the boredom part :) </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506397'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>added a comment.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506421'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>it would be an glitch in the system if the start and end are exactly the same so we are not interested on those.  </p><p> </p><p>Agree that the datetime/date is confusing, if the name of the variable in the original dataset, we made a comment about data in the data description side.</p><p> </p><p>Fixed the comment, that was wrong. </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506513'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>thank you :)</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506548'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>Yes, we were surprised as well. Either they are robust or the ids get recicled, we do mention that somewhere in the text. </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506573'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>good catch! removed.</p><p> </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506676'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>fixed.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506705'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>:)</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506737'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>fixed.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506760'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>we think that is TFL responding to the increased demand on bikes in early 2020, we've made a comment on the text about it.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506864'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>fixed.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506893'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>fixed</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506924'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>no reason, made it consistent.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517507009'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>adding a comment about the 30m, seems is still the case.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517709959'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>added a link!</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517709975'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>good catch! we've added copyright notice as required in the attribution requirements. </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1518179614'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>changed it to make sure runs for everyone</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1518179630'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>fixed.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506974'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>Thanks for the comments! The typo has been fixed.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517507041'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>Thank you both,  I agree is a bit confusing and there was some confusion in the code between chains and journeys. In this case each colour is a journey and the straight lines means that the bike was moved by TFL and not a cyclist so the chain ends and a new one starts. I have added this explanation to the notebook. </p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506992'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by crangelsmith** [comment]
<p>This is fixed now.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506621'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>Thanks for catching this. This turned out to be a bug that affected all the numbers in this table. It's now fixed in the code, just need to rerun the whole notebook once to see the effect.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506799'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>plt.show() added, thanks. Will look into the how the size of the plots will present in the final output.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506832'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>I increased the default figure size substantially. Will keep an eye on how it renders in the final version.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1517506832'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Review by llewelld — APPROVED** [review]
Thanks so much for addressing the comments so carefully. I read through again and it all looks great to me. I noted just a few typos which I've flagged up in the comments. Apart from these very minor changes, it looks good to go to me.

I also tested it running locally again (the commands I used are below in case you need to check my process is correct). Everthing worked first time; really great stuff.

Thanks again for a really good read!
```
cd stories/2022-10-06-Boris-Bikes
python3 -m venv venv
. ./venv/bin/activate
pip install pip --upgrade
pip install -r requirements.txt
pip install jupyterlab
pip install ipykernel
python3 -m ipykernel install --user --name venv-bikes
jupyter-lab
```

**Comment by mhauru** [comment]
<p>Done</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1621416742'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>Done</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1621416845'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>Whoops, good catch. Done.</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1621416921'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>Done</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1621416991'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by mhauru** [comment]
<p>Thanks for a great review!</p>

---

 <a href='https://app.reviewnb.com/alan-turing-institute/TuringDataStories/pull/221/discussion/#comment-1621417046'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Pull request merged**